### ISSUE TITLE
ClassNotFoundException, NoClassDefFoundError with Jackson Databind

### ISSUE THREAD

**Issue by grolschie** [vitalidze/chromecast-java-api-v2#118]
On the whole, _chromecast-java-api-v2_ is working very well for me. I have had to add exceptions to my Proguard rules file to exclude obfuscation of the entire library for this to work in release builds. Recently I have been receiving errors from Android Vitals.  I have no idea what the user input was unfortunately.

I have now excluded `com.fasterxml.jackson.databind` from my Proguard rules file and here is the result:

```
java.lang.NoClassDefFoundError: 
  at com.fasterxml.jackson.databind.ext.Java7HandlersImpl.getSerializerForJavaNioFilePath (Java7HandlersImpl.java)
  at com.fasterxml.jackson.databind.ext.OptionalHandlerFactory.findSerializer (OptionalHandlerFactory.java)
  at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findOptionalStdSerializer (BasicSerializerFactory.java)
  at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findSerializerByPrimaryType (BasicSerializerFactory.java)
  at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2 (BeanSerializerFactory.java)
  at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer (BeanSerializerFactory.java)
  at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer (SerializerProvider.java)
  at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer (SerializerProvider.java)
  at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer (SerializerProvider.java)
  at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer (SerializerProvider.java)
  at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue (DefaultSerializerProvider.java)
  at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue (ObjectMapper.java)
  at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString (ObjectMapper.java)
  at su.litvak.chromecast.api.v2.Channel.write (Channel.java)
  at su.litvak.chromecast.api.v2.Channel.access$000 (Channel.java)
  at su.litvak.chromecast.api.v2.Channel$PingThread.run (Channel.java)
  at su.litvak.chromecast.api.v2.Channel.connect (Channel.java)
  at su.litvak.chromecast.api.v2.Channel.open (Channel.java)
  at su.litvak.chromecast.api.v2.ChromeCast.connect (ChromeCast.java)
  at [myApp].CastFragment.access$600 (CastFragment.java)
  or                     .handlerTask (CastFragment.java)
  at [myApp].CastFragment.access$002 (CastFragment.java)
  or                     .access$1000 (CastFragment.java)
  or                     .access$102 (CastFragment.java)
  or                     .access$1300 (CastFragment.java)
  or                     .access$402 (CastFragment.java)
  or                     .access$500 (CastFragment.java)
  or                     .itemSend (CastFragment.java)
  or                     .onCreateView (CastFragment.java)
  or                     .removeItem (CastFragment.java)
  at [myApp].CastFragment$9.run (CastFragment.java)
  at java.lang.Thread.run (Thread.java:818)
  
Caused by: java.lang.ClassNotFoundException: 
  at dalvik.system.BaseDexClassLoader.findClass (BaseDexClassLoader.java:56)
  at java.lang.ClassLoader.loadClass (ClassLoader.java:511)
  at java.lang.ClassLoader.loadClass (ClassLoader.java:469)
  at com.fasterxml.jackson.databind.ext.Java7HandlersImpl.getSerializerForJavaNioFilePath (Java7HandlersImpl.java)
  at com.fasterxml.jackson.databind.ext.OptionalHandlerFactory.findSerializer (OptionalHandlerFactory.java)
  at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findOptionalStdSerializer (BasicSerializerFactory.java)
  at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findSerializerByPrimaryType (BasicSerializerFactory.java)
  at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2 (BeanSerializerFactory.java)
  at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer (BeanSerializerFactory.java)
  at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer (SerializerProvider.java)
  at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer (SerializerProvider.java)
  at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer (SerializerProvider.java)
  at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer (SerializerProvider.java)
  at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue (DefaultSerializerProvider.java)
  at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue (ObjectMapper.java)
  at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString (ObjectMapper.java)
  at su.litvak.chromecast.api.v2.Channel.write (Channel.java)
  at su.litvak.chromecast.api.v2.Channel.access$000 (Channel.java)
  at su.litvak.chromecast.api.v2.Channel$PingThread.run (Channel.java)
  at su.litvak.chromecast.api.v2.Channel.connect (Channel.java)
  at su.litvak.chromecast.api.v2.Channel.open (Channel.java)
  at su.litvak.chromecast.api.v2.ChromeCast.connect (ChromeCast.java)
  at [myApp].CastFragment.access$600 (CastFragment.java)
  or                     .handlerTask (CastFragment.java)
  at [myApp].CastFragment.access$002 (CastFragment.java)
  or                     .access$1000 (CastFragment.java)
  or                     .access$102 (CastFragment.java)
  or                     .access$1300 (CastFragment.java)
  or                     .access$402 (CastFragment.java)
  or                     .access$500 (CastFragment.java)
  or                     .itemSend (CastFragment.java)
  or                     .onCreateView (CastFragment.java)
  or                     .removeItem (CastFragment.java)
  at [myApp].CastFragment$9.run (CastFragment.java)
  at java.lang.Thread.run (Thread.java:818)

**Comment by grolschie** [comment]
Running the debug version of my app on a AVD running Android 6, API 23 "Marshmallow" crashes my app with:

```
10-16 12:25:22.880 [myApp].pro E/AndroidRuntime: FATAL EXCEPTION: Thread-201
    Process: [myApp].pro, PID: 4138
    java.lang.NoClassDefFoundError: Failed resolution of: Ljava/nio/file/Path;
        at com.fasterxml.jackson.databind.ext.Java7HandlersImpl.getSerializerForJavaNioFilePath(Java7HandlersImpl.java:28)
        at com.fasterxml.jackson.databind.ext.OptionalHandlerFactory.findSerializer(OptionalHandlerFactory.java:96)
        at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findOptionalStdSerializer(BasicSerializerFactory.java:439)
        at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findSerializerByPrimaryType(BasicSerializerFactory.java:374)
        at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:226)
        at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:165)
        at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1388)
        at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1336)
        at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:510)
        at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:713)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:308)
        at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:4094)
        at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:3404)
        at su.litvak.chromecast.api.v2.Channel.write(Channel.java:400)
        at su.litvak.chromecast.api.v2.Channel.access$000(Channel.java:51)
        at su.litvak.chromecast.api.v2.Channel$PingThread.run(Channel.java:123)
        at su.litvak.chromecast.api.v2.Channel.connect(Channel.java:323)
        at su.litvak.chromecast.api.v2.Channel.open(Channel.java:282)
        at su.litvak.chromecast.api.v2.ChromeCast.connect(ChromeCast.java:165)
        at [myApp].CastFragment.handlerTask(CastFragment.java:486)
        at [myApp].CastFragment.access$1300(CastFragment.java:44)
        at [myApp].CastFragment$9.run(CastFragment.java:422)
        at java.lang.Thread.run(Thread.java:818)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "java.nio.file.Path" on path: DexPathList[[zip file "/data/app/[myApp].pro-1/base.apk"],nativeLibraryDirectories=[/data/app/[myApp].pro-1/lib/x86, /vendor/lib, /system/lib]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
        at com.fasterxml.jackson.databind.ext.Java7HandlersImpl.getSerializerForJavaNioFilePath(Java7HandlersImpl.java:28) 
        at com.fasterxml.jackson.databind.ext.OptionalHandlerFactory.findSerializer(OptionalHandlerFactory.java:96) 
        at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findOptionalStdSerializer(BasicSerializerFactory.java:439) 
        at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findSerializerByPrimaryType(BasicSerializerFactory.java:374) 
        at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:226) 
        at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:165) 
        at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1388) 
        at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1336) 
        at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:510) 
        at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:713) 
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:308) 
        at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:4094) 
        at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:3404) 
        at su.litvak.chromecast.api.v2.Channel.write(Channel.java:400) 
        at su.litvak.chromecast.api.v2.Channel.access$000(Channel.java:51) 
        at su.litvak.chromecast.api.v2.Channel$PingThread.run(Channel.java:123) 
        at su.litvak.chromecast.api.v2.Channel.connect(Channel.java:323) 
        at su.litvak.chromecast.api.v2.Channel.open(Channel.java:282) 
        at su.litvak.chromecast.api.v2.ChromeCast.connect(ChromeCast.java:165) 
        at [myApp].CastFragment.handlerTask(CastFragment.java:486) 
        at [myApp].CastFragment.access$1300(CastFragment.java:44) 
        at [myApp].CastFragment$9.run(CastFragment.java:422) 
        at java.lang.Thread.run(Thread.java:818) 
    	Suppressed: java.lang.ClassNotFoundException: java.nio.file.Path
        at java.lang.Class.classForName(Native Method)
        at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
        at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
        		... 24 more
     Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available

**Comment by grolschie** [comment]
I'm suspecting an issue with Jackson Databind on certain versions of Android.  A similar issue was here:
[https://github.com/getsentry/sentry-java/issues/493](https://github.com/getsentry/sentry-java/issues/493)

Here's the doc page for the Java7Support class:
[https://fasterxml.github.io/jackson-databind/javadoc/2.9/com/fasterxml/jackson/databind/ext/Java7Support.html#getSerializerForJavaNioFilePath-java.lang.Class-](https://fasterxml.github.io/jackson-databind/javadoc/2.9/com/fasterxml/jackson/databind/ext/Java7Support.html#getSerializerForJavaNioFilePath-java.lang.Class-)

**Comment by grolschie** [comment]
Problem identified:
The problem does **not** occur with chromecast-java-api-v2 version `0.11.0` which downloads an earlier version of the Jackson libraries.

**Comment by grolschie** [comment]
I'm thinking that this is the problem and it appears to have been patched:
https://github.com/FasterXML/jackson-databind/issues/2466

**Comment by vitalidze** [comment]
Thank you for reporting and performing investigation. 

Since `0.11.1` the project has a wider set of jackson versions to include, which can download some beta versions prior to 2.10.0. However, you can define the specific version of jackson in your project. The library supports any version in the range `[2.8.11,2.10.0)`. For example you may use `2.9.10.1` that was released yesterday. Does it work for you?

**Comment by grolschie** [comment]
Please excuse my ignorance. Do you mean I that I should manually add an entry to my build.gradle file for a particular Jackson version, and that will cause your library to not download another? Or that I should somehow specify a particular Jackson version to your library?

**Comment by vitalidze** [comment]
It's the first one: I meant you can add a dependency on a particular Jackson version into your `build.gradle` file within range `[2.8.11,2.10.0)` (for example `2.9.10.1` will work). Then the project will use that version and you should not see the issue anymore.

**Comment by vitalidze** [comment]
BTW the change in jackson version was introduced in #115

**Comment by grolschie** [comment]
Thank you for that. Adding `implementation 'com.fasterxml.jackson.core:jackson-databind:2.10.0'` to my build.gradle file works. As does adding version `2.9.10.1` as you suggested.

Leaving it out of my build.gradle file still crashes though - so it needs to be added.

**Comment by vitalidze** [comment]
To be honest I am not sure how to provide a comprehensive solution for all cases here since the library itself tends to use stable versions of jackson and will work with 2.8, 2.9 and 2.10 (though the 2.10 stable I haven't tested myself yet). Maybe hard coding something like `[2.8.2,2.10.999.999]` might work in case there will be not versions greater than `2.10.999.999` (hopefully).

As for the `implementation` configuration instead of `compile` it applies for android applications. For an ordinary java application it would be the `compile` configuration. I will think about amending the README.md.

**Comment by grolschie** [comment]
Re: `compile `. Ahh sorry about that. :-)

Re: the Jackson dependencies. It is possible to somehow just add an exclusion for the version number with the bug?

**Comment by vitalidze** [comment]
I have just released a new version of the library (`0.11.3`) with the more strict dependencies on jackson. This should fix the issue.
